data/transactions/logic/frames.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opProto(cx *EvalContext) error {
+	if !cx.fromCallsub {
+		return fmt.Errorf("proto executed in normal flow")
+	}
+	cx.fromCallsub = false
+	nargs := int(cx.program[cx.pc+1])
+	if nargs > len(cx.stack) {
+		return fmt.Errorf("callsub to proto that requires %d args with stack height %d", nargs, len(cx.stack))
+	}
+	top := len(cx.callstack) - 1
+	cx.callstack[top].clear = true
+	cx.callstack[top].args = nargs
+	cx.callstack[top].returns = int(cx.program[cx.pc+2])
+	return nil
+}
+
+func opFrameGet(cx *EvalContext) error {
+	i := int8(cx.program[cx.pc+1])
+
+	top := len(cx.callstack) - 1
+	if top < 0 {
+		return errors.New("frame_get with empty callstack")
+	}
+
+	frame := cx.callstack[top]
+	if i < 0 && -int(i) > frame.args {
+		return fmt.Errorf("frame_get %d in sub with %d args", i, frame.args)
+	}
+	idx := frame.height + int(i)
+	if idx >= len(cx.stack) {
+		return errors.New("frame_get above stack")
+	}
+	if idx < 0 { // Impossible, because of arglen check in callsub
+		return errors.New("frame_get below stack")
+	}
+
+	cx.stack = append(cx.stack, cx.stack[idx])
+	return nil
+}
+func opFramePut(cx *EvalContext) error {
+	last := len(cx.stack) - 1 // value
+	i := int8(cx.program[cx.pc+1])
+
+	top := len(cx.callstack) - 1
+	if top < 0 {
+		return errors.New("frame_put with empty callstack")
+	}
+
+	frame := cx.callstack[top]
+	if i < 0 && -int(i) > frame.args {
+		return fmt.Errorf("frame_put %d in sub with %d args", i, frame.args)
+	}
+	idx := frame.height + int(i)
+	if idx >= last {
+		return errors.New("frame_put above stack")
+	}
+	if idx < 0 { // Impossible, because of arglen check in callsub
+		return errors.New("frame_put below stack")
+	}
+	cx.stack[idx] = cx.stack[last]
+	cx.stack = cx.stack[:last] // pop value
+	return nil
+}
+
+func opPushN(cx *EvalContext) error {
+	n := cx.program[cx.pc+1]
+	cx.stack = append(cx.stack, make([]stackValue, n)...)
+	return nil
+}
```

Comment:
`pushn` is at your service.

Reply:
`pushn` helps a lot, but on the type stack, it seems we are filling `n uint64`. Though I am hesitate to propose `n Any`, for it would lose all the information of type, which is not preferable, I don't see a better way than current `pushn` to preserve space on stack then. Otherwise one has to write bytes on stack, bury in some depth in the frame, I think.

Reply:
push n does push uint64s, so the type stack is correct after `puishn X`.  Now, your program may intend to use those slots as bytes, and that's ok, as long as you use `frame_bury` to push bytes there before you access them.  I used 0 as the default in the same way that scratch slots start out with 0 in them.

Reply:
I think that makes sense as a space allocator, should be enough to use.